File: src/lib/crypto.ts
/**
 * Small Web Crypto helpers used by the sync engine (file hashing) and the
 * S3-compatible client (AWS SigV4 signing). No native dependencies, so they
 * run identically in the browser and the Tauri webview.
 */

function toBytes(data: string | Uint8Array): Uint8Array {
  if (typeof data === "string") return new TextEncoder().encode(data);
  return data;
}

export async function sha256Hex(message: string | Uint8Array): Promise<string> {
  const bytes = toBytes(message);
  const ab = new ArrayBuffer(bytes.byteLength);
  new Uint8Array(ab).set(bytes);
  const digest = await crypto.subtle.digest("SHA-256", ab);
  return Array.from(new Uint8Array(digest))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

export async function hmacSha256(
  key: Uint8Array,
  message: string,
): Promise<Uint8Array> {
  const cryptoKey = await crypto.subtle.importKey(
    "raw",
    key as BufferSource,
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const sig = await crypto.subtle.sign(
    "HMAC",
    cryptoKey,
    new TextEncoder().encode(message) as BufferSource,
  );
  return new Uint8Array(sig);
}

export async function hmacSha256Hex(
  key: Uint8Array,
  message: string,
): Promise<string> {
  const sig = await hmacSha256(key, message);
  return Array.from(sig)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

File: Mobile2/my_diary/lib/sync/crypto.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:crypto/crypto.dart';

/// SHA-256 / HMAC-SHA256 助手（同步引擎文件指纹 + S3 SigV4 签名）。无原生依赖。
String sha256Hex(List<int> bytes) => sha256.convert(bytes).toString();

Uint8List hmacSha256(Uint8List key, String message) =>
    Uint8List.fromList(Hmac(sha256, key).convert(utf8.encode(message)).bytes);

/// 直接取 Digest 的 toString()（与 sha256Hex 一致，输出 64 位小写 hex）。
/// 注意不能对 Uint8List 调 toString()——它输出的是 "[217, 55, ...]" 这种
/// Dart 列表表示，会生成非法签名导致 S3/MinIO/OSS 全部拒签。
String hmacSha256Hex(Uint8List key, String message) =>
    Hmac(sha256, key).convert(utf8.encode(message)).toString();
